docs/design/specs/Memory-model.md

```diff
@@ -0,0 +1,298 @@
+
+# .NET memory model
+
+## ECMA 335 vs. .NET memory models
+ECMA 335 standard defines a very weak memory model. After two decades the desire to have a flexible model did not result in considerable benefits due to hardware being more strict. On the other hand programming against ECMA model requires extra complexity to handle scenarios that are hard to comprehend and not possible to test.
+
+In the course of multiple releases .NET runtime implementations settled around a memory model that is a practical compromise between what can be implemented efficiently on the current hardware, while staying reasonably approachable by the developers. This document rationalizes the invariants provided and expected by the .NET runtimes in their current implementation with expectation of that being carried to future releases.
+
+## Alignment
+When managed by the .NET runtime, variables of built-in primitive types are *properly aligned* according to the data type size. This applies to both heap and stack allocated memory.
+
+1-byte, 2-byte, 4-byte variables are stored at 1-byte, 2-byte, 4-byte boundary, respectively.
+8-byte variables are 8-byte aligned on 64 bit platforms.
+Native-sized integer types and pointers have alignment that matches their size on the given platform.
+
+The alignment of fields is not guaranteed when `FieldOffsetAttribute` is used to explicitly adjust field offsets.
+
+## Atomic memory accesses
+Memory accesses to *properly aligned* data of primitive and Enum types with size with sizes up to the platform pointer size are always atomic. The value that is observed is always a result of complete read and write operations.
+
+Primitive types: bool, char, int8, uint8, int16, uint16, int32, uint32, int64, uint64, float32, float64, native int, native unsigned int.
+
+Values of unmanaged pointers are treated as native integer primitive types. Memory accesses to *properly aligned* values of unmanaged pointers are atomic.
+
+Managed references are always aligned to their size on the given platform and accesses are atomic.
+
+The following methods perform atomic memory accesses regardless of the platform when the location of the variable is managed by the runtime.
+- `System.Threading.Interlocked` methods
+- `System.Threading.Volatile` methods
+
+**Example:** `Volatile.Read<double>(ref location)` on a 32 bit platform is atomic, while an ordinary read of `location` may not be.
```

Comment:
@VSadov does this section guarantee that e.g. `Volatile.Read<long>(ref location)` is atomic on 32bit? Because JIT doesn't do that currently, e.g.:
```csharp
long Foo(ref long location) => Volatile.Read(ref location);
```
arm32 codegen:
```asm
000002  6808           ldr     r0, [r1]
000004  F3BF 8F5F      dmb     15
000008  6849           ldr     r1, [r1+0x04]
00000A  F3BF 8F5F      dmb     15
```
we _probably can_ fix it actually via simd(fpu on x64) if it's needed.

Reply:
The official docs have stated that guarantee for a long time, and there's definitely code that relies on it.

Reply:
Filed https://github.com/dotnet/runtime/issues/81999

Reply:
> arm32 codegen

Is this actual codegen on arm32, or is it disasm produced via non-full fidelity codegen (aljit, pmi, etc.)? I do not see how we can end up with this codegen on arm32.

> we probably can fix it actually via simd(fpu on x64) if it's needed

I do not think that 64-bit simd or floating point operations on x86 and arm32 guarantee atomicity.

The BCL implementation provides this invariant by using 64-bit Interlocked operation for 32-bit: https://github.com/dotnet/runtime/blob/main/src/libraries/System.Private.CoreLib/src/System/Threading/Volatile.cs#L95

And the VM intrinsic expansion of the 64-bit volatile methods is disabled for 32-bit targets: https://github.com/dotnet/runtime/blob/main/src/coreclr/vm/jitinterface.cpp#L7064-L7066

Reply:
Thanks, sorry for wasted time -I should have tested it on a real 32bit jit instead